bpf: exclude pod's reply traffic from egress gateway logic

jibi opened the pull request:
Currently all pod traffic matching source IP and destination CIDR of an
egress policy will be forwarded to an egress gateway.

This means we will incorrectly forward to an egress gateway also all
reply traffic from connections destined to a pod, breaking said
connections.

This commit fixes this by adding an additional check to make sure reply
traffic (i.e. connections not originating from a pod) is excluded from
the egress gateway logic.

Fixes: #17866
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -803,6 +804,15 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
+		 * from a pod.
+		 */
+		if (!ct_is_reply4(get_ct_map4(&tuple), ctx, ETH_HLEN + ipv4_hdrlen(ip4), &tuple,
+				  &is_reply) && is_reply)
```
Can't we use `ret != CT_REPLY` instead of doing a new map lookup here?

jibi replied:
Ah, I think we can :+1: 

I got fooled by the `CT_EGRESS` param passed to the `ct_lookup4()` call above, but I just noticed that this will set `tuple->flags` to `TUPLE_F_IN` and so we'll end up with the same key set by `ct_is_reply4()`

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1167,29 +1182,27 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+				 * skipped forwarding the packet over the tunnel
+				 * to the egress gateway in bpf_lxc)
+				 */
+				if (!ct_is_reply) {
```
Shouldn't we check *before* the egress map lookup, to avoid an unnecessary lookup?

pchaigno commented on bpf/lib/nodeport.h:
```diff
@@ -1167,29 +1182,27 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+				 * skipped forwarding the packet over the tunnel
+				 * to the egress gateway in bpf_lxc)
+				 */
+				if (!ct_is_reply) {
```
If `ep` is NULL, `ct_is_reply` will always be false, even if it is actually reply traffic, because we will never execute `ct_is_reply4` above. Or am I missing something?

jibi replied:
We want to skip SNAT only if traffic is both coming from a local endpoint/pod and is reply traffic, so if the first condition is false (`ep == NULL`) we don't really care if it's reply traffic or not

pchaigno replied:
Ah, I get it. If it's coming from a remote node (i.e., `ep == NULL`), it can't be reply traffic anyway. I think the comment should clarify this as it's probably not going to be obvious to anyone who's not deeply familiar with the egress gateway.

pchaigno commented on test/k8sT/Egress.go:
```diff
@@ -149,6 +154,37 @@ var _ = SkipDescribeIf(func() bool {
 		res = kubectl.ExecPodCmd(randomNamespace, srcPod, "dig kubernetes +time=2")
 		res.ExpectSuccess()
+
+		if ciliumOpts["tunnel"] == "disabled" {
```
Why this check? Does the bug not affect tunneling mode?

pchaigno commented on test/k8sT/Egress.go:
```diff
@@ -149,6 +154,37 @@ var _ = SkipDescribeIf(func() bool {
+			outsideNodeName, outsideNodeIP := kubectl.GetNodeInfo(helpers.GetNodeWithoutCilium())
+
+			// Add a route for the target pod's IP on the node running without Cilium to
+			// allow reaching it from outside the cluster
```
Can't we expose the pod via a NodePort service instead? I'm guessing it's because we use a direct route that we can't run this in tunneling mode?

jibi replied:
> Can't we expose the pod via a NodePort service instead?

That's what I initially tried (https://github.com/cilium/cilium/commit/61eecd0123e548832348d27ec2ef801e0c6c893f) but that test was succeeding even without the reply traffic checks, since to my understanding:
* if the target pod is running on the same node to which we connect to, once we get a reply from the pod we will jump to a different program before even looking up the egress gw policy: https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L710-L714
* if the target pod is running on a different node, the node to which we connected to is going to SNAT the traffic with its own IP before forwarding it to the second node, and so even in this case the egress gw logic will be skipped since the reply traffic is destined to a remote node (https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L799)

> I'm guessing it's because we use a direct route that we can't run this in tunneling mode?

Yep, since we can't use a nodeport I run this test only in direct routing. This actually works even in tunneling mode (since `bpf_host` is attached anyway to the physical device), I'm just not sure it's worth having that test :thinking:

pchaigno replied:
Thanks for the explanation! It makes sense to me.

AFAICS, that means the additional check in the datapath is only necessary for native routing mode today. I'm tempted to limit the check to native routing, but... :point_down: 

> if the target pod is running on the same node to which we connect to, once we get a reply from the pod we will jump to a different program before even looking up the egress gw policy: https://github.com/cilium/cilium/blob/master/bpf/bpf_lxc.c#L710-L714

This one relies on the assumption that we're using BPF to perform service resolution. If we ever enable the egress gateway to work with kube-proxy instead of KPR and we're relying on this assumption (e.g., we limited the check to only native routing mode), then the egress gateway may silently break. So probably best not to rely on that assumption.

Considering this, do you think it would be worth testing both pod-to-service in addition to direct pod-to-pod?

pchaigno replied:
(Whatever you decide, I think it would be worth a comment on the `ciliumOpts["tunnel"] == "disabled"` check to clarify the assumptions made here.)

jibi replied:
> Considering this, do you think it would be worth testing both pod-to-service in addition to direct pod-to-pod?

Yep, I guess it doesn't hurt to have an additional test for `external node -> nodeport service` since it can detect regressions on this front, so I'll add it back both for direct routing and tunneling :+1:

brb commented on bpf/bpf_lxc.c:
```diff
@@ -803,6 +803,14 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 		if (!info)
 			goto skip_egress_gateway;
 
+		/* If the packet is a reply, it means that outside has initiated
```
Nit: maybe mention in the comment that we exclude related packets as well.

brb commented on bpf/lib/nodeport.h:
```diff
@@ -1144,7 +1145,28 @@ static __always_inline bool snat_v4_needed(struct __ctx_buff *ctx, __be32 *addr,
+
+			if (!ct_is_reply4(get_ct_map4(&tuple), ctx, ETH_HLEN + ipv4_hdrlen(ip4),
+					  &tuple, &is_reply) && is_reply)
+				ct_is_reply = true;
```
Why can't we just `return false` here instead? This would eliminate the `if(!ct_is_reply)` below.

jibi replied:
No strong reason here, I just thought avoiding this early return would make the code below a bit more explicit

brb replied:
I think it's in opposite - the early return improves the code readability (no need to keep the context in head until the end of the function), and reduces the instr count.

kkourt reviewed (approved):
🚀

brb commented on bpf/bpf_lxc.c:
```diff
@@ -605,7 +605,7 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 		return ret;
 
-	reason = ret;
+	ct_ret = ret;
```
I feel nit-picky today, but why can't we use `ct_ret` directly a few lines above, i.e., `ct_ret = ct_lookup4(...)`?

jibi replied:
Yep, no reason for not doing that :+1:

brb reviewed (approved):
One non-blocking nit. Besides that :rocket: :moon:

jibi committed:
bpf: rename reason to ct_ret in handle_ipv4_from_lxc

to make it more explicit its purpose

Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>

pchaigno commented on bpf/bpf_lxc.c:
```diff
@@ -525,7 +525,7 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 	__u8 __maybe_unused encrypt_key = 0;
 	__u32 monitor = 0;
-	__u8 reason;
+	__u8 ct_ret;
```
Should we also rename in other functions (IPv6, bpf_host, etc.) for consistency?

I'm also wondering if this change should be in its own PR to *not* backport it as it may cause a lot of conflicts.

jibi replied:
> Should we also rename in other functions (IPv6, bpf_host, etc.) for consistency?

Yes, I was thinking about doing it in a separate PR as it's a bit out of scope with the changes proposed here

> I'm also wondering if this change should be in its own PR to not backport it as it may cause a lot of conflicts.

I'll probably take care of backporting this PR anyway since there are conflicts also in the first commit, so I think it's fine

jibi committed:
bpf: exclude pod's reply traffic from egress gateway logic

Currently all pod traffic matching source IP and destination CIDR of an
egress policy will be forwarded to an egress gateway.

This means we will incorrectly forward to an egress gateway also all
reply traffic from connections destined to a pod, breaking said
connections.

This commit fixes this by adding an additional check to make sure reply
traffic (i.e. connections not originating from a pod) is excluded from
the egress gateway logic.

Fixes: #17866
Signed-off-by: Gilberto Bertin <gilberto@isovalent.com>
